### PR TITLE
Define test cases

### LINKED ISSUE #98: When must default answer files exist?
Honest question. The specification is silent about when a test case must provide `.ans`.

1. I’m reasonably sure that for `validation: default` it is true that if `<base_name>.in` exists then `<base_name>.ans` must also exist. We should say that.
2. I _think_ this is true for `validation: custom`. This is highly implicit, but since `judge_answer` in the specification needs a value, the adventurous reader can guess that this value was taken to be `<base_name>.ans`. I’m not sure, though.
3. The answer is not “always”, because explicitly now, for `validation: interactive: True` there need be no such file in `data/sample`

Is the rule “whenever an input file `.in` exists then a default answer file `.ans` must also exist”? (I think so.)

### PR DESCRIPTION
**Pull request by Tagl**
Hopefully closes #98

### CODE REVIEW

File: spec/2023-07-draft.md [hunk 1]
```diff
@@ -322,15 +328,24 @@ All files and folders associated to a single test case have the same base name w
...
+For that purpose, a judge system could choose to reuse the result of a previous run, or re-run the equivalent test case.
+
 ### Input
 
 Each test case can supply input via standard input, command-line arguments, and/or the file system. 
 These options are not exclusive. 
-For a test case with base name `test`, the file `test.in` (if it exists) is piped to the submission as standard input. 
-The submission's filename is passed as the first command-line argument, followed by the whitespace-separated tokens in `test.args` (if this file exists).
+For a test case with base name `test`, the file `test.in` is piped to the submission as standard input. 
+The submission's filename is passed as the first command-line argument, followed by the whitespace-separated tokens in `test.args`, if the file exists.
```
Comment (Tagl) [inline spec/2023-07-draft.md:344]: I do not think this is the right wording.
This is correct in the case of a single file submission and, for example, python.
In general this is not true and I think we should simply skip mentioning the first argument or say that it follows from how programs are run in general, so the main entry point is the first argument, whether it be binary, main class or main file.

File: spec/2023-07-draft.md [hunk 2]
```diff
@@ -322,15 +328,24 @@ All files and folders associated to a single test case have the same base name w
 | `.args`                         | Input                | Input passed as command-line arguments |
 | `.files`                        | Input                | Input available via file I/O           | 
 
+Judge systems may assume that the result of running a program on a test case is deterministic. 
+In other words, for any two test cases, if the contents of their `.in` and `.args` files,
```
Comment (niemela) [inline spec/2023-07-draft.md:332]:
```suggestion
For any two test cases, if the contents of their `.in` and `.args` files,
```

File: spec/2023-07-draft.md [hunk 3]
```diff
@@ -322,15 +328,24 @@ All files and folders associated to a single test case have the same base name w
...
+and `.files` folder are equivalent, then the input of the two test cases is equivalent. 
+Additionally, for any two test cases, if their input,
```
Comment (niemela) [inline spec/2023-07-draft.md:334]:
```suggestion
This means that for any two test cases, if their input,
```

File: spec/2023-07-draft.md [hunk 4]
```diff
@@ -322,15 +328,24 @@ All files and folders associated to a single test case have the same base name w
...
+<div class="not-icpc"> output validator flags (see test data groups below)</div>
+and the contents of their `.ans` files are equivalent, then the test cases are equivalent. 
+For that purpose, a judge system could choose to reuse the result of a previous run, or re-run the equivalent test case.
```
Comment (niemela) [inline spec/2023-07-draft.md:337]:
```suggestion
The assumption of determinism means that a judge system could choose to reuse the result of a previous run, or to re-run the equivalent test case.
```

File: spec/2023-07-draft.md (hunk 1 above)
Comment (niemela) [reply spec/2023-07-draft.md:344]: That's a good point. I really don't know how to write that though... :)

**Review by niemela — CHANGES_REQUESTED** [review]

File: spec/2023-07-draft.md (hunk 1 above)
Comment (Tagl) [reply spec/2023-07-draft.md:344]: Will make ticket for this

**Review by niemela — APPROVED** [review]

**Comment by niemela** [comment]
I'll merge this. @Tagl please create a ticket for the reminding issue that you brought up.

**Pull request merged**